xtask/src/flash_image.rs

```diff
@@ -190,85 +158,116 @@ impl FlashImage {
                 }
                 _ => return Err("Invalid image identifier")?,
             }
-
-            offset += 12;
         }
 
         println!("Image is valid!");
         Ok(())
     }
 }
 
-impl FlashImageHeader {
-    fn serialize(&self) -> Vec<u8> {
-        let mut buffer = Vec::new();
-        buffer.extend_from_slice(&self.magic_number.to_le_bytes());
-        buffer.extend_from_slice(&self.header_version.to_le_bytes());
-        buffer.extend_from_slice(&self.image_count.to_le_bytes());
-        buffer
-    }
+pub fn calculate_checksum(data: &[u8]) -> u32 {
+    let mut hasher = Hasher::new();
+    hasher.update(data);
+    hasher.finalize()
 }
 
-impl FlashImagePayload {
-    pub fn serialize(&self) -> Vec<u8> {
-        let mut buffer = Vec::new();
-        for info in &self.image_info {
-            buffer.extend_from_slice(&info.identifier.to_le_bytes());
-            buffer.extend_from_slice(&info.image_offset.to_le_bytes());
-            buffer.extend_from_slice(&info.size.to_le_bytes());
+impl FlashImagePayload<'_> {
+    pub fn calculate_checksum(&self) -> u32 {
+        let mut hasher = Hasher::new();
+        for info in self.image_info {
+            hasher.update(info.as_bytes());
         }
-        for image in &self.images {
-            buffer.extend_from_slice(&image.data);
+        for image in self.images {
+            hasher.update(image.data);
         }
-        buffer
+        hasher.finalize()
     }
 }
 
 impl FlashImageChecksum {
     pub fn new(header: &FlashImageHeader, payload: &FlashImagePayload) -> Self {
         Self {
-            header: Self::calculate_checksum(&header.serialize()),
-            payload: Self::calculate_checksum(&payload.serialize()),
+            header: calculate_checksum(header.as_bytes()),
+            payload: payload.calculate_checksum(),
         }
     }
-    pub fn calculate_checksum(data: &[u8]) -> u32 {
-        let mut hasher = Hasher::new();
-        hasher.update(data);
-        hasher.finalize()
-    }
+}
+
+fn load_file(filename: &str) -> Result<Vec<u8>, DynError> {
+    let mut buffer = Vec::new();
+
+    // Open the file, map errors to a custom error message
+    let mut file =
+        File::open(filename).map_err(|e| format!("Cannot open file '{}': {}", filename, e))?;
+
+    // Read the file into the buffer, map errors similarly
+    file.read_to_end(&mut buffer)
+        .map_err(|e| format!("Cannot read file '{}': {}", filename, e))?;
+
+    let padding = buffer.len().next_multiple_of(4) - buffer.len(); // Calculate padding size
+    buffer.extend(vec![0; padding]); // Append padding bytes
+
+    Ok(buffer)
 }
 
 pub(crate) fn flash_image_create(
     caliptra_fw_path: &str,
     soc_manifest_path: &str,
     mcu_runtime_path: &str,
-    soc_image_paths: &Option<Vec<String>>,
+    soc_image_paths: &Option<Vec<&str>>,
```

Comment:
I think it's okay if this is `&Option<Vec<String>>` or `&Option<Vec<PathBuf>>` so that we don't need to have the conversion function below.

Reply:
ack